[FLINK-37028] Datastream Window operator with async state processing

jiexray opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Implement DataStream WindowOperator to support async state access.


## Brief change log

TODO


## Verifying this change

- UT: `AsyncWindowOperatorTest`
- IT: `testAsyncWindowWordCount` & `AsyncEventTimeWindowCheckpointingITCase`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

Zakelly commented on flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/declare/state/AggregatingStateWithDeclaredNamespace.java:
```diff
@@ -0,0 +1,111 @@
+
+    @Override
+    public void mergeNamespaces(N target, Collection<N> sources) {
+        state.mergeNamespaces(target, sources);
+        state.mergeNamespaces(target, sources);
```
why twice...

jiexray replied:
Resolved.

Zakelly commented on flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/declare/state/StateWithDeclaredNamespace.java:
```diff
@@ -0,0 +1,84 @@
+        this.declaredNamespace = declaredNamespace;
+    }
+
+    protected void wrapNamespace() {
```
A little strange here... How about naming it `resetNamespace`

jiexray replied:
resolved

Zakelly reviewed (commented):
Thanks for the PR! I will post my comments incrementally.

Zakelly commented on flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/declare/state/StateWithDeclaredNamespace.java:
```diff
@@ -0,0 +1,84 @@
+
+/**
+ * State wrapped by declared namespace. This is useful for window operator, where we need to
+ * frequently setup current namespace in each lambda. This wrap class help use to invoke {@code
```
Suggestion:

A partitioned state that wraps a declared namespace and hide the namespace switching from user. User will use the state just like the public state APIs without any namespaces manipulation. Useful in window operators.

jiexray replied:
Resolved. 

Changed to:
```
/**
 * A partitioned state that wraps a declared namespace and hide the namespace switching from user. 
 * User will only use the state just like the public state APIs without any consideration on namespace. 
 * 
 * <p>This wrap class is useful in DataStream window operation, where namespace is managed by the operator
 * and user function is free from namespace manipulation.
 */
```

Zakelly commented on flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/declare/state/StateWithDeclaredNamespace.java:
```diff
@@ -0,0 +1,84 @@
+ * setCurrentNamespace} when we are going to call async state API.
+ */
+@Internal
+public class StateWithDeclaredNamespace<N> {
```
How about implement `InternalKeyedState` here

jiexray replied:
Good idea. I have changed it and let `StateWithDeclaredNamespace` unify the impl of `setCurrentNamespace(N)`

Zakelly commented on flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/declare/state/ValueStateWithDeclaredNamespace.java:
```diff
@@ -0,0 +1,75 @@
+
+    @Override
+    public void setCurrentNamespace(N namespace) {
+        state.setCurrentNamespace(namespace);
```
The declared namespace should also be set here. Then the `mergeNamespaces` will work properly.

jiexray replied:
resolved.

After let `StateWithDeclaredNamespace` implements `InternalKeyedState`, Base class `StateWithDeclaredNamespace` unifies the impl of `setCurrentNamespace ()` in  all XxxWithDeclaredNamespace class in:
```
    @Override
    public void setCurrentNamespace(N namespace) {
        declaredNamespace.set(namespace);
        state.setCurrentNamespace(namespace);
    }
```

Zakelly commented on flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/operators/windowing/functions/InternalAggregateProcessAsyncWindowFunction.java:
```diff
@@ -0,0 +1,114 @@
+
+        ProcessWindowFunction<V, R, K, W> wrappedFunction = this.wrappedFunction;
+        wrappedFunction.clear(ctx);
+        return StateFutureUtils.completedFuture(null);
```
nit. Replace all `StateFutureUtils.completedFuture(null);` to `StateFutureUtils.completedVoidFuture();`

jiexray replied:
Resolved

Zakelly commented on flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/operators/windowing/functions/InternalAggregateProcessAsyncWindowFunction.java:
```diff
@@ -0,0 +1,114 @@
+ * @param <V> The type of the AggregateFunction's result, and the input to the WindowFunction
+ * @param <R> The result type of the WindowFunction
+ */
+public final class InternalAggregateProcessAsyncWindowFunction<T, ACC, V, R, K, W extends Window>
```
It seems this function is not used for now?

Zakelly commented on flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/declare/state/StateWithDeclaredNamespace.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <N, S extends State> S stateWrapWithNamespace(
```
```suggestion
    public static <N, S extends State> S create(
```

jiexray replied:
resolved

Zakelly commented on flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/operators/windowing/AsyncWindowOperator.java:
```diff
@@ -0,0 +1,902 @@
+    protected transient WindowContext processContext;
+
+    protected transient DeclaredVariable<W> windowDeclaredVariable;
+    protected transient DeclaredVariable<K> keyDeclaredVariable;
```
Seems unnecessary

jiexray replied:
Resolved. I have removed it.

Zakelly commented on flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/operators/windowing/AsyncWindowOperator.java:
```diff
@@ -0,0 +1,902 @@
+    }
+
+    /** Internal class for keeping track of in-flight timers. */
+    protected static class Timer<K, W extends Window> implements Comparable<Timer<K, W>> {
```
Could this class be removed?

jiexray replied:
Resolved

Zakelly commented on flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/operators/windowing/AsyncWindowOperator.java:
```diff
@@ -0,0 +1,902 @@
+        }
+    }
+
+    protected <S extends State, N, T> S getAsyncPartitionedState(
+            N namespace, TypeSerializer<N> namespaceSerializer, StateDescriptor<T> stateDescriptor)
+            throws Exception {
+        return getAsyncKeyedStateBackend()
+                .getOrCreateKeyedState(namespace, namespaceSerializer, stateDescriptor);
+    }
```
I'd suggest this method includes `stateWrapWithNamespace`. WDYT?

And no need for `Async`, just `getPartitionedState ` seems good

jiexray replied:
Resolved, I just delete this useless function and let the caller directly use ` getAsyncKeyedStateBackend().getOrCreateKeyedState()`, See

```
Context#getPartitionedState(StateDescriptor<T> stateDescriptor)
```

Zakelly commented on flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/operators/windowing/AsyncWindowOperator.java:
```diff
@@ -0,0 +1,902 @@
+     * This is given to the {@code InternalWindowFunction} for emitting elements with a given
+     * timestamp.
+     */
+    protected transient TimestampedCollector<OUT> timestampedCollector;
```
Please use `TimestampedCollectorWithDeclaredVariable` instead.

jiexray replied:
Resolved

Zakelly commented on flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/operators/windowing/AsyncWindowOperator.java:
```diff
@@ -0,0 +1,902 @@
+        LOG.info("Initialize async window operator with trigger " + trigger);
+    }
+
+    @Override
+    public void setup(
+            StreamTask<?, ?> containingTask,
+            StreamConfig config,
+            Output<StreamRecord<OUT>> output) {
+        super.setup(containingTask, config, output);
+    }
+
+    @Override
+    public void setProcessingTimeService(ProcessingTimeService processingTimeService) {
+        super.setProcessingTimeService(processingTimeService);
+    }
```
Seems no need to override.

jiexray replied:
Resolved

Zakelly commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/windowing/assigners/WindowAssigner.java:
```diff
@@ -64,6 +65,12 @@ public abstract Collection<W> assignWindows(
     public abstract Trigger<T, W> getDefaultTrigger();
 
+    /** Returns the default async trigger associated with this {@code WindowAssigner}. */
+    public AsyncTrigger<T, W> getDefaultAsyncTrigger() {
```
Add `@Experimental` or `@internal`

jiexray replied:
WindowAssigner is stateless, I think it is not a must to introudce a async version.

Differently, Trigger is bind with state operation, and I need to introduce an async Trigger.

jiexray commented:
@Zakelly Thanks for reviewing. I have fixed the comments.

Moreover, I have added another two commits to run an end-to-end example (i.e., WindowWordCount) for window datastream with async state API.

Zakelly commented on flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/windowing/WindowWordCount.java:
```diff
@@ -118,15 +129,33 @@ public static void main(String[] args) throws Exception {
+                    keyedStream
+                            .enableAsyncState()
+                            // create windows of windowSize records slided every slideSize records
+                            .asyncCountWindow(windowSize, slideSize)
```
Ah.... we won't provide new API for user. Only `countWindow` combined with `enableAsyncState` is enough.

jiexray replied:
Resolved, I have removed this API. countWindow() + enableAsyncState() will trigger DataStream using AsyncWindowOperator.

Zakelly commented on flink-core/src/main/java/org/apache/flink/core/state/StateFutureUtils.java:
```diff
@@ -34,6 +35,14 @@
+     * Returns a function that always returns a completed future that does nothing and return null.
+     */
+    public static <T>
+            FunctionWithException<T, StateFuture<T>, Exception> createVoidReturnFunction() {
```
```suggestion
            FunctionWithException<T, StateFuture<Void>, Exception> returnVoidFuture() {
```

jiexray replied:
This util function is no use to me anymore, and I have removed it.

Zakelly commented on flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/operators/windowing/AsyncWindowOperator.java:
```diff
@@ -0,0 +1,810 @@
+                                                                    timer.getKey(),
+                                                                    timer.getNamespace(),
+                                                                    contents),
+                                                    StateFutureUtils.createVoidReturnFunction()),
```
Consider import static `StateFutureUtils.createVoidReturnFunction` since you are using this so many times.

Zakelly replied:
Or you can use the `thenConditionallyCompose` with only true branch provided.

jiexray replied:
Using `thenConditionallyCompose ` is a great idea, I have transformed to `thenConditionallyCompose` with only true branch.

 `StateFutureUtils.createVoidReturnFunction` has been removed.

Zakelly commented on flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/operators/windowing/AsyncWindowOperator.java:
```diff
@@ -0,0 +1,810 @@
+                                        ignore -> triggerResult.get().isPurge(),
+                                        ignore -> windowState.asyncClear(),
+                                        StateFutureUtils.createVoidReturnFunction())
+                                .thenApply(
+                                        empty -> {
+                                            registerCleanupTimer(window);
+                                            // if element is handled by none of assigned
+                                            // elementWindows
+                                            return false; // isSkippedElement = false
+                                        }));
```
```suggestion
                                .thenAccept(
                                        empty -> {
                                            registerCleanupTimer(window);
                                        }));
```

jiexray replied:
Thank you for your reminding. I have use `windowFutures` to implicitly indicate whether `isSkippedElement == true`, i.e., `windowFutures is empty => implicitly isSkippedElement == true`

Zakelly reviewed (commented):
LGTM. It's better to parameterize the test also.

Zakelly commented on flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/operators/windowing/triggers/AsyncTrigger.java:
```diff
@@ -0,0 +1,195 @@
+ * @param <T> The type of elements on which this {@code Trigger} works.
+ * @param <W> The type of {@link Window Windows} on which this {@code Trigger} can operate.
+ */
+@Experimental
```
We'd better mark all the async triggers as `@Internal` since they are not exposed to users.

jiexray replied:
resolved

Zakelly commented on flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/windowing/WindowWordCount.java:
```diff
@@ -88,6 +93,12 @@ public static void main(String[] args) throws Exception {
 
+        if (params.isAsyncState()) {
+            Configuration config = Configuration.fromMap(env.getConfiguration().toMap());
+            config.set(StateBackendOptions.STATE_BACKEND, FORST_STATE_BACKEND_NAME);
+            env.configure(config);
```
```suggestion
             if (!config.containsKey(StateBackendOptions.STATE_BACKEND.key())) {
                config.set(StateBackendOptions.STATE_BACKEND, FORST_STATE_BACKEND_NAME);
                env.configure(config);
            }
```
Same for the one in WordCount.java

jiexray replied:
Resolved, After change:
```
        // For async state, by default we will use the forst state backend.
        if (params.isAsyncState()) {
            Configuration config = Configuration.fromMap(env.getConfiguration().toMap());
            if (!config.containsKey(StateBackendOptions.STATE_BACKEND.key())) {
                config.set(StateBackendOptions.STATE_BACKEND, FORST_STATE_BACKEND_NAME);
                env.configure(config);
            }
        }
```

Zakelly commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/windowing/assigners/WindowAssigner.java:
```diff
@@ -64,6 +66,13 @@ public abstract Collection<W> assignWindows(
      */
     public abstract Trigger<T, W> getDefaultTrigger();
 
+    /** Returns the default async trigger associated with this {@code WindowAssigner}. */
+    @Experimental
+    public AsyncTrigger<T, W> getDefaultAsyncTrigger() {
+        // Subclass should override this method if they support async state processing.
+        throw new RuntimeException("The assigner does not support async state.");
+    }
+
```
Seems useless for now? Can we remove this?

jiexray replied:
Resolved. 

Yes, `getDefaultAsyncTrigger()` is no use, and I have removed it from `WindowAssigner`.

Zakelly commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/triggers/AsyncContinuousEventTimeTrigger.java:
```diff
@@ -0,0 +1,182 @@
+ * @see org.apache.flink.streaming.api.watermark.Watermark
+ */
+@PublicEvolving
+public class AsyncContinuousEventTimeTrigger<W extends Window> extends AsyncTrigger<Object, W> {
```
IIUC, this is not actually used. Can we move this to the package `org.apache.flink.runtime.asyncprocessing.operators.windowing.triggers` and mark as `@Internal`?

jiexray replied:
Resolved

jiexray commented:
> LGTM. It's better to parameterize the test also.

@Zakelly Thanks for reviewing. I have refactor the UT of window. After refactoring, existing window UT (i.e., `WindowOperatorTest` and `EvictingWindowOperatorTest`) can run with paramater `enableAsyncState`.

Zakelly reviewed (approved):
Thanks for the update! LGTM.

jiexray commented:
@flinkbot run azure

Zakelly commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/transformations/OneInputTransformation.java:
```diff
@@ -192,4 +192,9 @@ public boolean isOutputOnlyAfterEndOfStream() {
     public boolean isInternalSorterSupported() {
         return operatorFactory.getOperatorAttributes().isInternalSorterSupported();
     }
+
+    @Override
+    public void enableAsyncState() {
+        // nothing to do.
+    }
```
Here we should apply the change like [this PR](https://github.com/apache/flink/pull/26000/files#diff-a3035df99b00d9ab9a00e5050b2377d6cbb9ab7db0766ec9d88dc504a2643cfeR189-R198). Or after that is merged, rebase on that.
